File: sdks/go/client_test.go
package openagentpay

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// strptr is a small helper for optional string fields.
func strptr(s string) *string { return &s }

// newTestClient spins up an httptest server with the given handler and returns
// a Client pointed at it plus a cleanup func.
func newTestClient(t *testing.T, handler http.HandlerFunc, opts ...Option) (*Client, func()) {
	t.Helper()
	srv := httptest.NewServer(handler)
	c := NewClient(srv.URL, opts...)
	return c, srv.Close
}

func TestCreateSession(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/api/session" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		if ct := r.Header.Get("Content-Type"); ct != "application/json" {
			t.Errorf("expected JSON content-type, got %q", ct)
		}
		var body CreateSessionRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		if body.BudgetUsd != 10 || body.ExpiryMinutes != 30 {
			t.Errorf("unexpected body: %+v", body)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(SessionResponse{
			SessionID:     "sess_123",
			BudgetUsd:     10,
			ExpiryMinutes: 30,
			CreatedAt:     "2026-05-30T00:00:00Z",
			ExpiresAt:     "2026-05-30T00:30:00Z",
		})
	})
	defer cleanup()

	got, err := c.CreateSession(context.Background(), CreateSessionRequest{BudgetUsd: 10, ExpiryMinutes: 30})
	if err != nil {
		t.Fatalf("CreateSession error: %v", err)
	}
	if got.SessionID != "sess_123" || got.BudgetUsd != 10 || got.ExpiryMinutes != 30 {
		t.Errorf("unexpected session: %+v", got)
	}
}

func TestPaySuccess(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/api/pay" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body PayRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		if body.SessionID != "sess_123" || body.AmountUsdc != 1.5 {
			t.Errorf("unexpected pay body: %+v", body)
		}
		if body.Recipient == nil || *body.Recipient != "0xabc" {
			t.Errorf("expected recipient 0xabc, got %v", body.Recipient)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(PayResponse{
			Success:        true,
			TxHash:         strptr("0xdeadbeef"),
			ExplorerURL:    strptr("https://explorer/tx/0xdeadbeef"),
			AmountUsdc:     1.5,
			AmountAtomic:   "1500000",
			Payer:          "0xagent",
			Recipient:      "0xabc",
			Network:        "base-sepolia",
			WalletProvider: "hashkey",
			PaymentPayload: PaymentPayload{
				ChainID:           84532,
				VerifyingContract: "0xUSDC",
				Authorization: PaymentAuthorization{
					From: "0xagent", To: "0xabc", Value: "1500000",
					ValidAfter: "0", ValidBefore: "9999999999", Nonce: "0x01",
				},
				Signature: "0xsig", V: 27, R: "0xr", S: "0xs",
			},
		})
	})
	defer cleanup()

	got, err := c.Pay(context.Background(), PayRequest{
		SessionID:  "sess_123",
		AmountUsdc: 1.5,
		Recipient:  strptr("0xabc"),
	})
	if err != nil {
		t.Fatalf("Pay error: %v", err)
	}
	if !got.Success || got.TxHash == nil || *got.TxHash != "0xdeadbeef" {
		t.Errorf("unexpected pay response: %+v", got)
	}
	if got.PaymentPayload.ChainID != 84532 || got.PaymentPayload.Authorization.Value != "1500000" {
		t.Errorf("unexpected payload: %+v", got.PaymentPayload)
	}
}

func TestPayErrorStatuses(t *testing.T) {
	cases := []struct {
		name        string
		status      int
		body        string
		wantCode    string
		wantMessage string
	}{
		{
			name:        "402 payment required with error field",
			status:      http.StatusPaymentRequired,
			body:        `{"error":"insufficient_budget"}`,
			wantCode:    "insufficient_budget",
			wantMessage: "insufficient_budget",
		},
		{
			name:        "400 bad request with message + errorCode",
			status:      http.StatusBadRequest,
			body:        `{"message":"missing sessionId","errorCode":"bad_request"}`,
			wantCode:    "bad_request",
			wantMessage: "missing sessionId",
		},
		{
			name:        "500 non-json body",
			status:      http.StatusInternalServerError,
			body:        "internal boom",
			wantCode:    "http_500",
			wantMessage: "internal boom",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
				w.WriteHeader(tc.status)
				_, _ = io.WriteString(w, tc.body)
			})
			defer cleanup()

			_, err := c.Pay(context.Background(), PayRequest{SessionID: "x", AmountUsdc: 1})
			if err == nil {
				t.Fatal("expected error, got nil")
			}
			var apiErr *APIError
			if !errors.As(err, &apiErr) {
				t.Fatalf("expected *APIError, got %T: %v", err, err)
			}
			if apiErr.Status != tc.status {
				t.Errorf("status: got %d want %d", apiErr.Status, tc.status)
			}
			if apiErr.Code != tc.wantCode {
				t.Errorf("code: got %q want %q", apiErr.Code, tc.wantCode)
			}
			if apiErr.Message != tc.wantMessage {
				t.Errorf("message: got %q want %q", apiErr.Message, tc.wantMessage)
			}
			if apiErr.Raw != tc.body {
				t.Errorf("raw: got %q want %q", apiErr.Raw, tc.body)
			}
		})
	}
}

func TestGetSessionFound(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/session/sess_abc" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(SessionResponse{SessionID: "sess_abc", BudgetUsd: 5})
	})
	defer cleanup()

	got, err := c.GetSession(context.Background(), "sess_abc")
	if err != nil {
		t.Fatalf("GetSession error: %v", err)
	}
	if got.SessionID != "sess_abc" {
		t.Errorf("unexpected session id: %s", got.SessionID)
	}
}

func TestGetSession404(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_, _ = io.WriteString(w, `{"error":"session_not_found"}`)
	})
	defer cleanup()

	_, err := c.GetSession(context.Background(), "missing")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *APIError, got %T", err)
	}
	if apiErr.Status != http.StatusNotFound {
		t.Errorf("status: got %d want 404", apiErr.Status)
	}
	if apiErr.Code != "session_not_found" {
		t.Errorf("code: got %q want session_not_found", apiErr.Code)
	}
}

func TestGetSessionEscapesID(t *testing.T) {
	var gotPath string
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.EscapedPath()
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(SessionResponse{SessionID: "weird/id"})
	})
	defer cleanup()

	if _, err := c.GetSession(context.Background(), "weird/id"); err != nil {
		t.Fatalf("GetSession error: %v", err)
	}
	if !strings.Contains(gotPath, "weird%2Fid") {
		t.Errorf("expected escaped id in path, got %q", gotPath)
	}
}

func TestListWallets(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/api/wallets" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(WalletsResponse{
			Wallets: []WalletInfo{{
				WalletProvider: "hashkey",
				DisplayName:    "HashKey",
				ChainName:      "Base Sepolia",
				ChainID:        84532,
				TokenLabel:     "USDC",
				TokenAddress:   "0xUSDC",
				AgentAddress:   "0xagent",
			}},
			DefaultProvider: "hashkey",
		})
	})
	defer cleanup()

	got, err := c.ListWallets(context.Background())
	if err != nil {
		t.Fatalf("ListWallets error: %v", err)
	}
	if got.DefaultProvider != "hashkey" || len(got.Wallets) != 1 {
		t.Errorf("unexpected wallets: %+v", got)
	}
	if got.Wallets[0].ChainID != 84532 {
		t.Errorf("unexpected chainId: %d", got.Wallets[0].ChainID)
	}
}

func TestGetGovernanceAndAudit(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch r.URL.Path {
		case "/api/governance":
			_, _ = io.WriteString(w, `{"maxPerTx":100,"enabled":true}`)
		case "/api/governance/audit":
			_, _ = io.WriteString(w, `{"events":[{"id":"e1"}]}`)
		default:
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
	})
	defer cleanup()

	gov, err := c.GetGovernance(context.Background())
	if err != nil {
		t.Fatalf("GetGovernance error: %v", err)
	}
	if gov["enabled"] != true {
		t.Errorf("unexpected governance: %+v", gov)
	}

	audit, err := c.GetAudit(context.Background())
	if err != nil {
		t.Fatalf("GetAudit error: %v", err)
	}
	if _, ok := audit["events"]; !ok {
		t.Errorf("expected events key, got %+v", audit)
	}
}

func TestPayOnce(t *testing.T) {
	var sawExpiry int
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch r.URL.Path {
		case "/api/session":
			var body CreateSessionRequest
			_ = json.NewDecoder(r.Body).Decode(&body)
			sawExpiry = body.ExpiryMinutes
			_ = json.NewEncoder(w).Encode(SessionResponse{SessionID: "sess_once", BudgetUsd: body.BudgetUsd})
		case "/api/pay":
			var body PayRequest
			_ = json.NewDecoder(r.Body).Decode(&body)
			if body.SessionID != "sess_once" {
				t.Errorf("pay used wrong session: %s", body.SessionID)
			}
			_ = json.NewEncoder(w).Encode(PayResponse{Success: true, AmountUsdc: body.AmountUsdc, AmountAtomic: "2000000"})
		default:
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
	})
	defer cleanup()

	got, err := c.PayOnce(context.Background(), PayOnceRequest{BudgetUsd: 20, AmountUsdc: 2})
	if err != nil {
		t.Fatalf("PayOnce error: %v", err)
	}
	if got.Session.SessionID != "sess_once" || !got.Payment.Success {
		t.Errorf("unexpected payOnce result: %+v", got)
	}
	if sawExpiry != 60 {
		t.Errorf("expected default expiry 60, got %d", sawExpiry)
	}
}

func TestPayOnceCustomExpiry(t *testing.T) {
	var sawExpiry int
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch r.URL.Path {
		case "/api/session":
			var body CreateSessionRequest
			_ = json.NewDecoder(r.Body).Decode(&body)
			sawExpiry = body.ExpiryMinutes
			_ = json.NewEncoder(w).Encode(SessionResponse{SessionID: "s2"})
		case "/api/pay":
			_ = json.NewEncoder(w).Encode(PayResponse{Success: true})
		}
	})
	defer cleanup()

	if _, err := c.PayOnce(context.Background(), PayOnceRequest{BudgetUsd: 5, AmountUsdc: 1, ExpiryMinutes: 15}); err != nil {
		t.Fatalf("PayOnce error: %v", err)
	}
	if sawExpiry != 15 {
		t.Errorf("expected expiry 15, got %d", sawExpiry)
	}
}

func TestAPIKeyHeader(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Authorization"); got != "Bearer secret-key" {
			t.Errorf("expected bearer header, got %q", got)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(WalletsResponse{DefaultProvider: "hashkey"})
	}, WithAPIKey("secret-key"))
	defer cleanup()

	if _, err := c.ListWallets(context.Background()); err != nil {
		t.Fatalf("ListWallets error: %v", err)
	}
}

func TestNoAPIKeyHeaderWhenUnset(t *testing.T) {
	c, cleanup := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Authorization"); got != "" {
			t.Errorf("expected no auth header, got %q", got)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(WalletsResponse{DefaultProvider: "hashkey"})
	})
	defer cleanup()

	if _, err := c.ListWallets(context.Background()); err != nil {
		t.Fatalf("ListWallets error: %v", err)
	}
}

func TestTrailingSlashBaseURL(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/wallets" {
			t.Errorf("expected clean path /api/wallets, got %q", r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(WalletsResponse{DefaultProvider: "hashkey"})
	}))
	defer srv.Close()

	// Append multiple trailing slashes; client must normalize them away.
	c := NewClient(srv.URL + "///")
	if _, err := c.ListWallets(context.Background()); err != nil {
		t.Fatalf("ListWallets error: %v", err)
	}
}

func TestNewClientDefaultsBaseURL(t *testing.T) {
	c := NewClient("")
	if c.baseURL != DefaultBaseURL {
		t.Errorf("expected default base url, got %q", c.baseURL)
	}
}

func TestWithHTTPClient(t *testing.T) {
	custom := &http.Client{}
	c := NewClient("https://example.com", WithHTTPClient(custom))
	if c.httpClient != custom {
		t.Error("expected injected http client to be used")
	}
	// nil injection must not clobber the default.
	c2 := NewClient("https://example.com", WithHTTPClient(nil))
	if c2.httpClient == nil {
		t.Error("expected non-nil http client after nil injection")
	}
}

func TestAPIErrorImplementsError(t *testing.T) {
	var err error = &APIError{Status: 402, Code: "x", Message: "nope"}
	if !strings.Contains(err.Error(), "402") || !strings.Contains(err.Error(), "nope") {
		t.Errorf("unexpected error string: %s", err.Error())
	}
}

File: packages/sdk/tests/client.test.ts
/**
 * @openagentpay/sdk — client tests
 *
 * All tests run against an injected fake `fetch` — no network, fully
 * deterministic. We assert on the request the client made (url, method,
 * headers, body) AND on how it decodes responses / errors.
 *
 * @license Apache-2.0
 */

import { describe, it, expect } from "vitest";
import {
  OpenAgentPayClient,
  OpenAgentPayApiError,
  type FetchLike,
} from "../src/index.js";
import type {
  SessionResponse,
  PayResponse,
  WalletsResponse,
} from "../src/index.js";

// ----------------------------------------------------------------------------
//  Fake fetch helpers
// ----------------------------------------------------------------------------

interface RecordedCall {
  url: string;
  method: string;
  headers: Record<string, string>;
  body: unknown;
}

interface FakeResponseSpec {
  status?: number;
  json?: unknown;
  text?: string;
}

/** Build a fake fetch that records calls and returns scripted responses. */
function makeFakeFetch(
  responder: (call: RecordedCall) => FakeResponseSpec
): { fetchImpl: FetchLike; calls: RecordedCall[] } {
  const calls: RecordedCall[] = [];
  const fetchImpl: FetchLike = async (input, init) => {
    const url = typeof input === "string" ? input : input.toString();
    const headersRecord: Record<string, string> = {};
    const rawHeaders = (init?.headers ?? {}) as Record<string, string>;
    for (const [k, v] of Object.entries(rawHeaders)) {
      headersRecord[k.toLowerCase()] = v;
    }
    let body: unknown;
    if (typeof init?.body === "string") {
      try {
        body = JSON.parse(init.body);
      } catch {
        body = init.body;
      }
    }
    const call: RecordedCall = {
      url,
      method: init?.method ?? "GET",
      headers: headersRecord,
      body,
    };
    calls.push(call);

    const spec = responder(call);
    const status = spec.status ?? 200;
    const text =
      spec.text !== undefined
        ? spec.text
        : spec.json !== undefined
        ? JSON.stringify(spec.json)
        : "";
    // Minimal Response-ish object covering what the client touches.
    return {
      ok: status >= 200 && status < 300,
      status,
      text: async () => text,
    } as unknown as Response;
  };
  return { fetchImpl, calls };
}

const BASE = "https://example.test";

const SESSION_FIXTURE: SessionResponse = {
  sessionId: "sess_123",
  budgetUsd: 5,
  expiryMinutes: 60,
  createdAt: "2026-05-30T00:00:00.000Z",
  expiresAt: "2026-05-30T01:00:00.000Z",
};

const PAY_FIXTURE: PayResponse = {
  success: true,
  txHash: "0xabc",
  explorerUrl: "https://sepolia.basescan.org/tx/0xabc",
  amountUsdc: 0.01,
  amountAtomic: "10000",
  payer: "0xPayer",
  recipient: "0xRecipient",
  network: "base-sepolia",
  walletProvider: "hashkey",
  paymentPayload: {
    chainId: 84532,
    verifyingContract: "0xUSDC",
    authorization: {
      from: "0xPayer",
      to: "0xRecipient",
      value: "10000",
      validAfter: "0",
      validBefore: "9999999999",
      nonce: "0xnonce",
    },
    signature: "0xsig",
    v: 27,
    r: "0xr",
    s: "0xs",
  },
};

const WALLETS_FIXTURE: WalletsResponse = {
  wallets: [
    {
      walletProvider: "hashkey",
      displayName: "HashKey",
      chainName: "HashKey Chain Testnet",
      chainId: 133,
      tokenLabel: "USDC",
      tokenAddress: "0xUSDC",
      agentAddress: "0xAgent",
    },
  ],
  defaultProvider: "hashkey",
};

// ----------------------------------------------------------------------------
//  Tests
// ----------------------------------------------------------------------------

describe("OpenAgentPayClient", () => {
  it("createSession POSTs to /api/session with json body", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: SESSION_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const res = await client.createSession({ budgetUsd: 5, expiryMinutes: 60 });

    expect(res).toEqual(SESSION_FIXTURE);
    expect(calls).toHaveLength(1);
    expect(calls[0]!.url).toBe(`${BASE}/api/session`);
    expect(calls[0]!.method).toBe("POST");
    expect(calls[0]!.headers["content-type"]).toBe("application/json");
    expect(calls[0]!.body).toEqual({ budgetUsd: 5, expiryMinutes: 60 });
  });

  it("pay POSTs to /api/pay and returns a success PayResponse", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: PAY_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const res = await client.pay({ sessionId: "sess_123", amountUsdc: 0.01 });

    expect(res.success).toBe(true);
    expect(res.txHash).toBe("0xabc");
    expect(res.paymentPayload.chainId).toBe(84532);
    expect(calls[0]!.url).toBe(`${BASE}/api/pay`);
    expect(calls[0]!.body).toEqual({ sessionId: "sess_123", amountUsdc: 0.01 });
  });

  it("pay includes recipient + walletProvider only when provided", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: PAY_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    await client.pay({
      sessionId: "sess_123",
      amountUsdc: 0.5,
      recipient: "0xDest",
      walletProvider: "circle",
    });

    expect(calls[0]!.body).toEqual({
      sessionId: "sess_123",
      amountUsdc: 0.5,
      recipient: "0xDest",
      walletProvider: "circle",
    });
  });

  it("pay omits undefined optional fields from the wire body", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: PAY_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    await client.pay({ sessionId: "sess_x", amountUsdc: 1 });

    const body = calls[0]!.body as Record<string, unknown>;
    expect("recipient" in body).toBe(false);
    expect("walletProvider" in body).toBe(false);
  });

  it("pay surfaces a 402 (payment required) as OpenAgentPayApiError", async () => {
    const { fetchImpl } = makeFakeFetch(() => ({
      status: 402,
      json: { error: "budget_exceeded", message: "over budget" },
    }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    await expect(
      client.pay({ sessionId: "sess_123", amountUsdc: 999 })
    ).rejects.toBeInstanceOf(OpenAgentPayApiError);

    try {
      await client.pay({ sessionId: "sess_123", amountUsdc: 999 });
      expect.unreachable("should have thrown");
    } catch (e) {
      const err = e as OpenAgentPayApiError;
      expect(err.status).toBe(402);
      expect(err.code).toBe("budget_exceeded");
      expect(err.message).toBe("budget_exceeded");
      expect(err.raw).toEqual({ error: "budget_exceeded", message: "over budget" });
    }
  });

  it("pay surfaces a 400 (bad request) using the message field", async () => {
    const { fetchImpl } = makeFakeFetch(() => ({
      status: 400,
      json: { message: "amountUsdc must be positive" },
    }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    try {
      await client.pay({ sessionId: "sess_123", amountUsdc: -1 });
      expect.unreachable("should have thrown");
    } catch (e) {
      const err = e as OpenAgentPayApiError;
      expect(err.status).toBe(400);
      expect(err.message).toBe("amountUsdc must be positive");
      expect(err.code).toBe("http_400");
    }
  });

  it("getSession GETs /api/session/:id and returns the session", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: SESSION_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const res = await client.getSession("sess_123");

    expect(res).toEqual(SESSION_FIXTURE);
    expect(calls[0]!.url).toBe(`${BASE}/api/session/sess_123`);
    expect(calls[0]!.method).toBe("GET");
    // GET must not set a content-type or body.
    expect(calls[0]!.headers["content-type"]).toBeUndefined();
    expect(calls[0]!.body).toBeUndefined();
  });

  it("getSession url-encodes the id", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: SESSION_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    await client.getSession("a/b c");

    expect(calls[0]!.url).toBe(`${BASE}/api/session/a%2Fb%20c`);
  });

  it("getSession throws OpenAgentPayApiError on 404", async () => {
    const { fetchImpl } = makeFakeFetch(() => ({
      status: 404,
      json: { error: "session_not_found" },
    }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    try {
      await client.getSession("missing");
      expect.unreachable("should have thrown");
    } catch (e) {
      const err = e as OpenAgentPayApiError;
      expect(err).toBeInstanceOf(OpenAgentPayApiError);
      expect(err.status).toBe(404);
      expect(err.code).toBe("session_not_found");
    }
  });

  it("listWallets GETs /api/wallets", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: WALLETS_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const res = await client.listWallets();

    expect(res.defaultProvider).toBe("hashkey");
    expect(res.wallets[0]!.chainId).toBe(133);
    expect(calls[0]!.url).toBe(`${BASE}/api/wallets`);
    expect(calls[0]!.method).toBe("GET");
  });

  it("getGovernance GETs /api/governance", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({
      json: { mode: "advisory", maxPerTx: 10 },
    }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const res = await client.getGovernance();

    expect(res).toEqual({ mode: "advisory", maxPerTx: 10 });
    expect(calls[0]!.url).toBe(`${BASE}/api/governance`);
  });

  it("getAudit GETs /api/governance/audit", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({
      json: { events: [{ type: "settlement.completed" }] },
    }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const res = await client.getAudit();

    expect((res as { events: unknown[] }).events).toHaveLength(1);
    expect(calls[0]!.url).toBe(`${BASE}/api/governance/audit`);
  });

  it("payOnce creates a session then pays under it (two calls, threaded sessionId)", async () => {
    const { fetchImpl, calls } = makeFakeFetch((call) =>
      call.url.endsWith("/api/session")
        ? { json: SESSION_FIXTURE }
        : { json: PAY_FIXTURE }
    );
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    const { session, payment } = await client.payOnce({
      budgetUsd: 5,
      amountUsdc: 0.01,
    });

    expect(session).toEqual(SESSION_FIXTURE);
    expect(payment.success).toBe(true);
    expect(calls).toHaveLength(2);
    expect(calls[0]!.url).toBe(`${BASE}/api/session`);
    expect(calls[0]!.body).toEqual({ budgetUsd: 5, expiryMinutes: 60 });
    expect(calls[1]!.url).toBe(`${BASE}/api/pay`);
    expect((calls[1]!.body as Record<string, unknown>)["sessionId"]).toBe(
      "sess_123"
    );
  });

  it("payOnce respects a custom expiryMinutes and forwards recipient", async () => {
    const { fetchImpl, calls } = makeFakeFetch((call) =>
      call.url.endsWith("/api/session")
        ? { json: SESSION_FIXTURE }
        : { json: PAY_FIXTURE }
    );
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    await client.payOnce({
      budgetUsd: 20,
      amountUsdc: 2,
      expiryMinutes: 15,
      recipient: "0xMerchant",
    });

    expect(calls[0]!.body).toEqual({ budgetUsd: 20, expiryMinutes: 15 });
    expect((calls[1]!.body as Record<string, unknown>)["recipient"]).toBe(
      "0xMerchant"
    );
  });

  it("sets Authorization: Bearer header when apiKey is provided", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: WALLETS_FIXTURE }));
    const client = new OpenAgentPayClient({
      baseUrl: BASE,
      apiKey: "oap_sk_test",
      fetchImpl,
    });

    await client.listWallets();

    expect(calls[0]!.headers["authorization"]).toBe("Bearer oap_sk_test");
  });

  it("omits Authorization header when no apiKey is provided", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: WALLETS_FIXTURE }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    await client.listWallets();

    expect(calls[0]!.headers["authorization"]).toBeUndefined();
  });

  it("strips trailing slashes from baseUrl so paths don't double up", async () => {
    const { fetchImpl, calls } = makeFakeFetch(() => ({ json: WALLETS_FIXTURE }));
    const client = new OpenAgentPayClient({
      baseUrl: `${BASE}///`,
      fetchImpl,
    });

    await client.listWallets();

    expect(calls[0]!.url).toBe(`${BASE}/api/wallets`);
  });

  it("falls back to a generic message on a non-JSON error body", async () => {
    const { fetchImpl } = makeFakeFetch(() => ({
      status: 500,
      text: "Internal Server Error",
    }));
    const client = new OpenAgentPayClient({ baseUrl: BASE, fetchImpl });

    try {
      await client.listWallets();
      expect.unreachable("should have thrown");
    } catch (e) {
      const err = e as OpenAgentPayApiError;
      expect(err.status).toBe(500);
      expect(err.message).toBe("Internal Server Error");
      expect(err.code).toBe("http_500");
    }
  });

  it("throws when constructed without baseUrl", () => {
    expect(
      () =>
        new OpenAgentPayClient({
          baseUrl: "",
          fetchImpl: (async () => new Response()) as unknown as FetchLike,
        })
    ).toThrow(/baseUrl/);
  });
});
